### PR TITLE
py/objint_*: Convert back to small int after binary op.

### PR DESCRIPTION
**Pull request by jimmo**
Inspired by https://github.com/micropython/micropython/pull/9516#discussion_r988475628 -- this behaviour was a surprise to me.

Before this change, long/mpz ints propagated into all future calculations.

For example, a relatively common operation like `x = a * b // c` where a,b,c all small ints would always result in a long/mpz int, even if it didn't need to, and then this would impact all future calculations with x.

This adds +24 bytes on PYBV11 but avoids heap allocations and potential surprises (e.g. `big-big` is now a small `0`, and can safely be accessed with MP_OBJ_SMALL_INT_VALUE).

There are several places in the `int_big*` tests that result in small integers (especially zero). I don't think this change makes those tests less valid -- it still correctly checks that they evaluate to the right thing. What we're missing is a way to check that you can print a small-valued mpz (as all these tests will now be going via small int printing).

The longlong implementation doesn't have tests, but I adapted the new int_big_to_small.py to run on a 32-bit Linux build.

_This work was funded through GitHub Sponsors._

### CODE REVIEW

**Comment by jimmo** [comment]
The coverage results are pretty much as expected:
 - mpz_not_inpl (no longer tests ~0)
 - mpz_as_bytes (fill rest of buffer with sign-ext or zero)
 - mp_obj_int_get_checked no longer ever hits the "is actually a small int" path
 - mp_obj_get_int_maybe (always small int)
 - mp_obj_id (always small int)

I guess this can all be addressed through `coverage.c`. @dpgeorge if you want to go ahead with this PR, then I can do that.

**Comment by dpgeorge** [comment]
Well... maybe we can delete all that code that's no longer "reachable"...?  If we can't write tests for it in Python, it's not strictly needed.

**Comment by dpgeorge** [comment]
Also would be worth running the performance benchmarks, to see if they change.

**Comment by jimmo** [comment]
> Well... maybe we can delete all that code that's no longer "reachable"...? If we can't write tests for it in Python, it's not strictly needed.

You make a very good point. I'm not sure if I'm brave enough to remove them though and it is a very small number of lines.

I realised that pow3 can be used to make "small big ints" still, but we should probably also add the same logic there too.

**Comment by dpgeorge** [comment]
I ran benchmark tests on this PR, on PYBV10:
```
diff of scores (higher is better)
N=100 M=100                  baseline ->    this-PR         diff      diff% (error%)
bm_chaos.py                    354.08 ->     355.34 :      +1.26 =  +0.356% (+/-0.03%)
bm_fannkuch.py                  73.86 ->      73.45 :      -0.41 =  -0.555% (+/-0.00%)
bm_fft.py                     2308.99 ->    2299.43 :      -9.56 =  -0.414% (+/-0.00%)
bm_float.py                   5735.70 ->    5753.49 :     +17.79 =  +0.310% (+/-0.01%)
bm_hexiom.py                    46.49 ->      46.53 :      +0.04 =  +0.086% (+/-0.00%)
bm_nqueens.py                 4222.19 ->    4202.58 :     -19.61 =  -0.464% (+/-0.00%)
bm_pidigits.py                 623.05 ->     676.10 :     +53.05 =  +8.515% (+/-0.21%)
bm_wordcount.py                 46.95 ->      46.47 :      -0.48 =  -1.022% (+/-0.00%)
core_import_mpy_multi.py       599.15 ->     599.81 :      +0.66 =  +0.110% (+/-0.01%)
core_import_mpy_single.py       97.62 ->      97.71 :      +0.09 =  +0.092% (+/-0.01%)
core_locals.py                  41.47 ->      41.13 :      -0.34 =  -0.820% (+/-0.00%)
core_qstr.py                   201.69 ->     202.16 :      +0.47 =  +0.233% (+/-0.00%)
core_str.py                     27.28 ->      27.34 :      +0.06 =  +0.220% (+/-0.00%)
core_yield_from.py             359.09 ->     358.79 :      -0.30 =  -0.084% (+/-0.00%)
misc_aes.py                    399.52 ->     404.30 :      +4.78 =  +1.196% (+/-0.00%)
misc_mandel.py                2973.05 ->    2991.61 :     +18.56 =  +0.624% (+/-0.01%)
misc_pystone.py               2348.74 ->    2371.03 :     +22.29 =  +0.949% (+/-0.00%)
misc_raytrace.py               369.76 ->     371.55 :      +1.79 =  +0.484% (+/-0.01%)
viper_call0.py                 503.99 ->     503.97 :      -0.02 =  -0.004% (+/-0.01%)
viper_call1a.py                527.68 ->     527.68 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1b.py                380.63 ->     380.63 :      +0.00 =  +0.000% (+/-0.00%)
viper_call1c.py                384.99 ->     384.98 :      -0.01 =  -0.003% (+/-0.01%)
viper_call2a.py                514.71 ->     514.73 :      +0.02 =  +0.004% (+/-0.01%)
viper_call2b.py                335.73 ->     335.73 :      +0.00 =  +0.000% (+/-0.00%)
```
The main test is `bm_pidigits.py` which makes heavy use of big-integer arithmetic.  And that test runs fasater!  I guess because there's now less memory allocations, and smaller integers are used where possible which are faster.

**Comment by dpgeorge** [comment]
> The coverage results are pretty much as expected
> ...
> I guess this can all be addressed through coverage.c

Visiting this PR 20 months later, there's a lot less uncovered code now.  There are only two small blocks that are now uncovered due to this PR: in `mpz_not_inpl()` when the argument is zero, and `mp_obj_int_get_checked()` when the argument is an mpz which is a small-int.  It's not worth removing these small bits of code, in case they are needed in the future.  So they have been left as-is and unix coverage tests added for them.

File: py/objint_mpz.c [hunk 1]
```diff
@@ -308,6 +308,13 @@ mp_obj_t mp_obj_int_binary_op(mp_binary_op_t op, mp_obj_t lhs_in, mp_obj_t rhs_i
             }
         }
 
+        mp_int_t res_small;
+        if (mpz_as_int_checked(&res->mpz, &res_small)) {
```
[the remark this answers is not in the document]
Comment (dpgeorge) [reply py/objint_mpz.c:317]: Performance testing has been done, see below.

**Pull request merged**